awx/main/tasks.py

```diff
@@ -149,6 +149,85 @@ def handle_setting_changes(self, setting_keys):
+    removed_queues = []
+    for queue in worker_queues:
+        if queue['name'] in settings.AWX_CELERY_QUEUES_STATIC or \
+                {'alias': queue['name']} in settings.AWX_CELERY_QUEUES_STATIC:
+            continue
+
+        removed_queues.append(queue['name'])
+        app.control.cancel_consumer(queue['name'], reply=True, destination=[worker_name])
+    return removed_queues
+
+
+def _add_queues(app, instance, worker_name):
+    # Add worker to group-named queues
+    queues = []
+    for ig in instance.rampart_groups.all():
+        app.control.add_consumer(ig.name, reply=True, destination=[worker_name])
+        queues.append(ig.name)
+
+    # Add worker to worker-named queue
+    app.control.add_consumer(instance.hostname, reply=True, destination=[worker_name])
+    queues.append(instance.hostname)
+
+    return queues
+
+
+def _update_celery_routes(instance):
+    logger.debug("Updating celery routes for instance {}".format(instance.hostname))
+
+    tasks = [
+        'awx.main.tasks.cluster_node_heartbeat',
+        'awx.main.tasks.purge_old_stdout_files',
+    ]
+
+    # Instance is, effectively, a controller node
+    if instance.rampart_groups.filter(controller__isnull=False).exists():
+        tasks.append('awx.main.tasks.awx_isolated_heartbeat')
+
+    for t in tasks:
+        settings.CELERY_ROUTES[t] = {'queue': instance.hostname, 'routing_key': instance.hostname}
+
+
+def _handle_ha_toplogy_changes():
+    instance = Instance.objects.me()
+    celery_worker_name = 'celery@%s' % instance.hostname
+
+    removed_queues = []
+    added_queues = []
+
+    _update_celery_routes(instance)
+
+    # Try catch this
+    app = Celery(broker=settings.BROKER_URL)
+    celery_host_queues = app.control.inspect([celery_worker_name]).active_queues()
+    if celery_host_queues is not None:
+        if celery_worker_name in celery_host_queues:
+            removed_queues = _cancel_queues(app, celery_host_queues[celery_worker_name], celery_worker_name)
+        else:
+            logger.warn('Local worker {} not found in celery queue list {}'.format(celery_worker_name, celery_host_queues))
+
+    added_queues = _add_queues(app, instance, celery_worker_name)
+    return (instance, removed_queues, added_queues)
+
+
+@task(queue='tower_broadcast_all', bind=True, base=LogErrorsTask)
+def handle_ha_toplogy_changes(self):
+    logger.debug("Reconfigure celeryd queues task")
+    (instance, removed_queues, added_queues) = _handle_ha_toplogy_changes()
+    logger.info("Workers on node '{}' removed from queues {} and added to queues {}"
+                .format(instance.hostname, removed_queues, added_queues))
+
+
+@worker_ready.connect
+def handle_ha_toplogy_worker_ready(sender, **kwargs):
+    logger.debug("Configure celeryd queues task")
+    (instance, removed_queues, added_queues) = _handle_ha_toplogy_changes()
+    logger.info("Workers on node '{}' added queues {}".format(instance.hostname, added_queues))
```

Comment:
Seems like it would be cleaner to move the log message into `_handle_ha_toplogy_changes`. I also think it would be nice to have separate messages for the added queues + removed queues, conditionally based on whether anything was changed. If something was changed, I would strongly argue in favor of INFO level of higher, so that they are emitted in production. It might be better as WARNING. If things are working normally (at steady-state), 0 should be added and 0 should be removed, so we just shouldn't be writing the log message in those cases.

Reply:
After reading more of the code, I realize that the information about which queues were changed might not actually be available.